src/storage/src/sink/kafka.rs

```diff
@@ -296,15 +300,39 @@ impl TransactionalProducer {
             .create_with_context(storage_configuration, ctx, &options, InTask::Yes)
             .await?;
 
+        // The partition count is fixed up after we ensure the topic exists.
+        let partition_count = Arc::new(AtomicU64::new(0));
+        let update_partition_count = {
+            let partition_count = Arc::clone(&partition_count);
+            let metrics = Arc::clone(&metrics);
+            Arc::new(move |pc| {
+                partition_count.store(pc, std::sync::atomic::Ordering::SeqCst);
+                metrics.partition_count.set(pc);
+            })
+        };
+
+        // Start a task that will keep the partition count up to date in the
+        // background.
+        let partition_count_task = task::spawn(
```

Comment:
do we need some sort of mechanism to cancel this task if the sink dataflow is stopped?

Reply:
We have that! We use an `AbortOnDrop` handle to automatically drop this task when the `TransactionalProducer` is dropped.